src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/ApplyBQSRDataflow.java

```diff
@@ -0,0 +1,153 @@
+        //SAMFileHeader header = readsSource.getHeader();
+        //final SAMSequenceDictionary sequenceDictionary = header.getSequenceDictionary();
+        //final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(sequenceDictionary) :
+        //        IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+        final BaseRecalOutput recalInfo = new BaseRecalOutput(BQSR_RECAL_FILE);
+        PCollection<BaseRecalOutput> recalInfoSingletonCollection = pipeline.apply(Create.of(recalInfo).withName("recal_file ingest"));
+        PCollection<Read> output = input // readsSource.getReadPCollection(intervals, ValidationStringency.SILENT)
+           .apply(new ApplyBQSRTransform(header, recalInfoSingletonCollection, bqsrOpts));
+        SmallBamWriter.writeToFile(pipeline, output, header, OUTPUT);
+    }
+
+    /** reads local disks or GCS -> header, and PCollection */
+    private PCollection<Read> ingestReadsAndGrabHeader(final Pipeline pipeline, String filename) throws IOException {
+
+        // input reads
+        if (BucketUtils.isCloudStorageUrl(filename)) {
+            // set up ingestion on the cloud
+            // but read the header locally
+            GcsPath path = GcsPath.fromUri(filename);
+            InputStream inputstream = Channels.newInputStream(new GcsUtil.GcsUtilFactory().create(pipeline.getOptions())
+                    .open(path));
+            SamReader reader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT).open(SamInputResource.of(inputstream));
+            header = reader.getFileHeader();
+
+            final SAMSequenceDictionary sequenceDictionary = header.getSequenceDictionary();
+            final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(sequenceDictionary) :
+                    IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
+            return new ReadsSource(filename, pipeline).getReadPCollection(intervals, ValidationStringency.SILENT);
+        } else {
+            // ingestion from local file
+            SamReader reader = SamReaderFactory.makeDefault().validationStringency(ValidationStringency.SILENT).open(new File(filename));
+            header = reader.getFileHeader();
+            List<Read> readLst = new ArrayList<>();
+            for (SAMRecord sr : reader) {
+                try {
+                    Read e = ReadConverter.makeRead(sr);
+                    readLst.add(e);
+                } catch (SAMException x) {
+                    logger.warn("Skipping read " + sr.getReadName() + " because we can't convert it. "+x.getMessage());
+                } catch (NullPointerException y) {
+                    logger.warn("Skipping read " + sr.getReadName() + " because we can't convert it. (null?)");
+                }
```

Comment:
This is essentially identical to a block of code above, and should probably be extracted from both into it's own function.